Draft fix for the OADP 4289 - to be tested before proposed upstream

mpryc opened the pull request:
Draft pull request
http://quay.io/migi/velero:fix_oadp_4289

dymurray committed:
Dockerfile.ubi/travis local files

add UBI dockerfiles
Use numeric user for velero-restic-restore-helper
Enable multiarch builds (#135)
Use arm64-graviton2 for arm builds (#137)
Add required keys for arm builds (#139)
Update Travis build job to work w/o changes on new branches
Use a full VM for arm
Use numeric non-root user for nonroot SCC compatibility

rayfordj committed:
Add BZ + Publish automation to repo (#82)

(cherry picked from commit ccb545fbcf293e0e7595a91763481cf4fa7e0d45)

Update PR-BZ automation mapping (#84)

(cherry picked from commit aa2b019366b98b64c46244e0a697540e7c4d5d2b)

Update PR-BZ automation (#92)

Co-authored-by: Rayford Johnson <rjohnson@redhat.com>
(cherry picked from commit ecc563f90324d9924318618f7a4a167f6ea231bf)

Add publish workflow (#108)

(cherry picked from commit f87b779eb243fa5b37731761613ea962103a5956)

sseago committed:
remove dependabot config from fork

kaovilai committed:
Create Makefile.prow

Code-gen no longer required on verify

due to https://github.com/vmware-tanzu/velero/pull/6039

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

oadp-1.2: Update Makefile.prow to velero-restore-helper

sseago committed:
set HOME in velero image for kopia, update controller-gen for CI (#280)

Signed-off-by: Scott Seago <sseago@redhat.com>

sseago committed:
build velero-helper binary for datamover pod

allenxu404 committed:
Ping golang/distroless image to latest version (#6679)

Signed-off-by: allenxu404 <qix2@vmware.com>

sseago committed:
Perf improvements for existing resource restore

Use informer cache with dynamic client for Get calls on restore
When enabled, also make the Get call before create.

Add server and install parameter to allow disabling this feature,
but enable by default

Signed-off-by: Scott Seago <sseago@redhat.com>

reasonerjt committed:
skip subresource in resource discovery (#6688)

Signed-off-by: lou <alex1988@outlook.com>
Co-authored-by: lou <alex1988@outlook.com>

Lyndon-Li committed:
fix issue 6647

Signed-off-by: Lyndon-Li <lyonghui@vmware.com>

kaovilai committed:
changelog

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

Lyndon-Li committed:
fix issue 6753

Signed-off-by: Lyndon-Li <lyonghui@vmware.com>

Commit:
Fix #6752: add namespace exclude check.

Add PSA audit and warn labels.

Signed-off-by: Xun Jiang <jxun@vmware.com>

ywk253100 committed:
Update restore controller logic for restore deletion (#6761)

1. Skip deleting the restore files from storage if the backup/BSL is not found
2. Allow deleting the restore files from storage even though the BSL is readonly

Signed-off-by: Wenkai Yin(尹文开) <yinw@vmware.com>

Lyndon-Li committed:
add csi snapshot data movement doc

Signed-off-by: Lyndon-Li <lyonghui@vmware.com>

Lyndon-Li committed:
issue 6786:always delete VSC regardless of the deletion policy

Signed-off-by: Lyndon-Li <lyonghui@vmware.com>

sseago committed:
issue #6807: Retry failed create when using generateName

When creating resources with generateName, apimachinery
does not guarantee uniqueness when it appends the random
suffix to the generateName stub, so if it fails with
already exists error, we need to retry.

Signed-off-by: Scott Seago <sseago@redhat.com>

allenxu404 committed:
Modify changelogs for v1.12

Signed-off-by: allenxu404 <qix2@vmware.com>

Lyndon-Li committed:
issue: move plugin depdending podvolume functions to util pkg

Signed-off-by: Lyndon-Li <lyonghui@vmware.com>

ywk253100 committed:
Replace the base image with paketobuildpacks image

Replace the base image with paketobuildpacks image

Fixes #6851

Signed-off-by: Wenkai Yin(尹文开) <yinw@vmware.com>

Lyndon-Li committed:
issue 6880: set ParallelUploadAboveSize as MaxInt64

Signed-off-by: Lyndon-Li <lyonghui@vmware.com>

allenxu404 committed:
Add doc links for new features to release note

Signed-off-by: allenxu404 <qix2@vmware.com>

dzaninovic committed:
Add support for block volumes (#6680) (#6897)

(cherry picked from commit 8e01d1b9be31d30f88e90003af6c4ce15c64ee65)

Signed-off-by: David Zaninovic <dzaninovic@catalogicsoftware.com>

Lyndon-Li committed:
issue 6734: spread backup pod evenly

Signed-off-by: Lyndon-Li <lyonghui@vmware.com>

0x113 committed:
Import auth provider plugins

Signed-off-by: Sebastian Glab <sglab@catalogicsoftware.com>

allenxu404 committed:
Add v1.12.1 changelog

Signed-off-by: allenxu404 <qix2@vmware.com>

Commit:
Make Windows build skip BlockMode code.

PVC block mode backup and restore introduced some OS specific
system calls. Those calls are not available for Windows, so
add both non Windows version and Windows version code, and
return error for block mode on the Windows platform.

Signed-off-by: Xun Jiang <jxun@vmware.com>

Lyndon-Li committed:
udmrepo use region specified in BSL when s3URL is empty

Signed-off-by: Lyndon-Li <lyonghui@vmware.com>

allenxu404 committed:
Change v1.12.1 changelog

Signed-off-by: allenxu404 <qix2@vmware.com>

kaovilai committed:
kopia/repository/config/aws.go: Set session.Options profile from config

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

kaovilai committed:
restore: Use warning when Create IsAlreadyExist and Get error

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

sseago committed:
use ubi9-latest to build

sseago committed:
OADP-4225: add tzdata to Dockerfile.ubi

Commit:
fix: CI

Signed-off-by: Mateus Oliveira <msouzaol@redhat.com>

sseago committed:
release-string for 1.14.0 (#318)

mpryc committed:
Draft fix for the OADP 4289 - to be tested before proposed upstream

Signed-off-by: Michal Pryc <mpryc@redhat.com>

mpryc commented:
@sseago @shubham-pampattiwar @weshayutin if you could comment if it's proper fix.

sseago commented on pkg/restore/restore.go:
```diff
@@ -1677,15 +1678,51 @@ func (ctx *restoreContext) restoreItem(obj *unstructured.Unstructured, groupReso
-				return warnings, errs, itemExists
+		err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			// First attempt to patch the object
+			_, err = resourceClient.Patch(name, patchBytes)
```
The first time we run this, we want to patch based on initial patchBytes which comes from the create call itself, as you have here. This is the normal case with no conflict, so we don't generate any extra APIServer calls. If it fails with Conflict, we grab obj again and retry. So for the first iteration, this is exactly what we want. However, the second iteration (assuming the grabbed obj is also a conflict) retries with the last iteration's patchBytes which we already know will result in a conflict. I'm thinking that we replace line 1685 and 1686 with:
```
if patchBytes != nil {
    _, err = resourceClient.Patch(name, patchBytes)
}
if patchBytes == nil || err != nil {
```

mpryc replied:
Makes sense. Changed.

mpryc replied:
https://github.com/openshift/velero/pull/331/commits/0e2f3392b702f4734847ba3db339b37ee726886f

sseago commented on pkg/restore/restore.go:
```diff
@@ -1677,15 +1678,51 @@ func (ctx *restoreContext) restoreItem(obj *unstructured.Unstructured, groupReso
+					if patchBytes != nil {
+						// Retry the patch operation with the updated patch bytes
+						_, err = resourceClient.Patch(name, patchBytes)
+						if err != nil {
```
After this line, set `patchBytes = nil` so that on the next retry we don't try to patch with a known-conflict patch.

mpryc replied:
Ah missed that, you are correct that this will try to patch with not-nil on a second attempt:

```go
			// First attempt to patch the object
			_, err = resourceClient.Patch(name, patchBytes)
```

sseago commented on pkg/restore/restore.go:
```diff
@@ -1677,15 +1678,51 @@ func (ctx *restoreContext) restoreItem(obj *unstructured.Unstructured, groupReso
+						_, err = resourceClient.Patch(name, patchBytes)
+						if err != nil {
+							return err
+						}
```
Add the log here so that post-conflict successful patches are logged just like those that succeed the first time:
```
	ctx.log.Infof("managed fields for %s patched successfully", kube.NamespaceAndName(obj))
```

mpryc committed:
Modifications after Scott suggestions.

Signed-off-by: Michal Pryc <mpryc@redhat.com>

kaovilai committed:
oadp-1.4: OADP-3227: Reconcile to fail on restore stuck in-progress (#330)

* oadp-1.4: OADP-3227: Mark InProgress backup/restore as failed upon requeuing (#315)

* Mark InProgress backup/restore as failed upon requeuing

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

remove uuid, return err to requeue instead of requeue: true

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

* cleanup to minimize diff from upstream

Signed-off-by: Scott Seago <sseago@redhat.com>

* error message update

Signed-off-by: Scott Seago <sseago@redhat.com>

* requeue on finalize status update.

Unlike the InProgress transition, there's no need to fail here,
since the Finalize steps can be repeated.

* Only run patch once for all backup finalizer return scenarios

Signed-off-by: Scott Seago <sseago@redhat.com>

---------

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>
Signed-off-by: Scott Seago <sseago@redhat.com>
Co-authored-by: Scott Seago <sseago@redhat.com>

* oadp-1.4: OADP-3227: Reconcile To Fail: Add backup/restore trackers (#324)

* OADP-4265: Reconcile To Fail: Add backup/restore trackers

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

* Apply suggestions from code review: backupTracker

* Address restoreTracker feedback

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

* s/delete from/add to/ in the comment

* unit test fix

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

* backup_controller unit test

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

* restore_controller unit test

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

* `make update`

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

* mock patch to fail failure due to connection refused

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

---------

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

* regenerate mocks

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>

---------

Signed-off-by: Tiger Kaovilai <tkaovila@redhat.com>
Signed-off-by: Scott Seago <sseago@redhat.com>
Co-authored-by: Scott Seago <sseago@redhat.com>

sseago commented on pkg/restore/restore.go:
```diff
@@ -1677,15 +1678,55 @@ func (ctx *restoreContext) restoreItem(obj *unstructured.Unstructured, groupReso
-			ctx.log.Infof("the managed fields for %s is patched", kube.NamespaceAndName(obj))
+			if patchBytes == nil || err != nil {
+				ctx.log.Errorf("error patching managed fields %s: %v", kube.NamespaceAndName(obj), err)
+				if apierrors.IsConflict(err) {
```
On the first execution of this func, Patch above may return a conflict, so this test is true -- and we regenerate and reapply patch. However, on subsequent executions, patchBytes is nil, so we don't have an error here, and therefore `IsConflict` is false. I think this should be `if err == nil || apierrors.IsConflict(err)`

mpryc replied:
Won't the `RetryOnConflict` make this only on `apierrors.IsConflict(err)` already?

sseago replied:
It will retry if that was the prior error. I'm not 100% certain that the value of the var `err` will be set here, though. In any case, we know that any time this func is  called, if `err` is nil at this point, that means we didn't just patch above, so we want to enter this block, so I think it's clearer if we enter any `err` is nil or is a conflict.

sseago commented on pkg/restore/restore.go:
```diff
@@ -1677,15 +1678,55 @@ func (ctx *restoreContext) restoreItem(obj *unstructured.Unstructured, groupReso
-		} else {
-			ctx.log.Infof("the managed fields for %s is patched", kube.NamespaceAndName(obj))
+			if patchBytes == nil || err != nil {
+				ctx.log.Errorf("error patching managed fields %s: %v", kube.NamespaceAndName(obj), err)
```
The above error log is appropriate if we just called `Patch` above, but not necessarily if `patchBytes` is nil. We could put another `if` block around it right here, or we could move it to after line 1686 above with that if block.

mpryc committed:
Revisited function for patching managed fields.

Signed-off-by: Michal Pryc <mpryc@redhat.com>

mpryc commented:
@sseago I've revisited that block of code and reworked it to be much simpler without so many if-else statements. Also checking if the error is type of conflict is unnecessary as the function `RetryOnConflict` won't retry on other errors.

If you could comment on current implementation.

sseago commented on pkg/restore/restore.go:
```diff
@@ -1669,24 +1670,48 @@ func (ctx *restoreContext) restoreItem(obj *unstructured.Unstructured, groupReso
+			withoutManagedFields = createdObj.DeepCopy()
+		}
+
+		createdObj.SetManagedFields(obj.GetManagedFields())
```
First time through, createdObj after this line is guaranteed to be exactly the same as withoutManagedFields except for the desired managed fields. But the second time through, createdObj is based on the original created object, which had the conflict error when we patched it. That means that this and withoutManagedFields may have other differences too.

sseago replied:
I'm thinking we want to use 2 local vars here -- withoutManagedFields and withManagedFields. Define it in the same if block above where we defined withoutManagedFields.

sseago replied:
See suggestions above, and for this line, s/createdObj/withManagedFields/

sseago replied:
No change to 1692 now. stick with createdObj.

sseago commented on pkg/restore/restore.go:
```diff
@@ -1669,24 +1670,48 @@ func (ctx *restoreContext) restoreItem(obj *unstructured.Unstructured, groupReso
+	var firstPatchTry bool = true
+
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		var withoutManagedFields *unstructured.Unstructured
```
var withoutManagedFields, withManagedFields

sseago replied:
Ignore above comment, no additional var needed.

sseago commented on pkg/restore/restore.go:
```diff
@@ -1669,24 +1670,48 @@ func (ctx *restoreContext) restoreItem(obj *unstructured.Unstructured, groupReso
+			if patchErr != nil {
+				ctx.log.Errorf("error fetching latest object for %s: %v", kube.NamespaceAndName(obj), patchErr)
+				return patchErr
 			}
```
withManagedFields = withoutManagedFields.DeepCopy()

sseago replied:
Skip this change too.

sseago commented on pkg/restore/restore.go:
```diff
@@ -1669,24 +1670,48 @@ func (ctx *restoreContext) restoreItem(obj *unstructured.Unstructured, groupReso
+
 		} else {
-			ctx.log.Infof("the managed fields for %s is patched", kube.NamespaceAndName(obj))
+			withoutManagedFields = createdObj.DeepCopy()
```
withManagedFields = createdObj
(I don't think we need deepcopy here, since this is the last use of createdObj)

mpryc replied:
The createdObj is used later, so I wonder if we need to do some assignment so createdObj becomes updated with managed fields in a situation we get one fresh from cluster ?

sseago replied:
Oh. Actually, in that case, a simpler change. Ignore the suggestion to create `withManagedFields` and instead, above we can set `createdObj` to the result of `resourceClient.Get` and omit the else clause completely. We'll do the deepCopy outside of the if block.

mpryc replied:
Should be fixed.

sseago commented on pkg/restore/restore.go:
```diff
@@ -1669,24 +1670,48 @@ func (ctx *restoreContext) restoreItem(obj *unstructured.Unstructured, groupReso
+		}
+
+		createdObj.SetManagedFields(obj.GetManagedFields())
+		patchBytes, patchErr = generatePatch(withoutManagedFields, createdObj)
```
s/createdObj/withManagedFields/

sseago replied:
No change here.

sseago commented on pkg/restore/restore.go:
```diff
@@ -1669,24 +1670,48 @@ func (ctx *restoreContext) restoreItem(obj *unstructured.Unstructured, groupReso
+			return patchErr
 		}
+		if patchBytes != nil {
+			// Retry the patch operation with the updated patch bytes
```
Minor rewording here, since this isn't always a retry now (same code used for initial try and retries). Or maybe just remove the comment, since it's clear from the code that we're patching now.

sseago commented on pkg/restore/restore.go:
```diff
@@ -1669,24 +1670,48 @@ func (ctx *restoreContext) restoreItem(obj *unstructured.Unstructured, groupReso
+
+	if err != nil {
+		ctx.log.Errorf("error patching managed fields %s: %v", kube.NamespaceAndName(obj), err)
+		errs.Add(namespace, err)
```
I don't think we need the `log.Errorf` here, since we already logged the error inside the retry func. Also, we need to add the if block around `errs.Add` and `return` here, since we don't error out if it's a NotFound error:
```
	if !apierrors.IsNotFound(err) {
		errs.Add(namespace, err)
		return warnings, errs, itemExists
	}
```

mpryc replied:
For that I think we need to also include:

```
	if err != nil && !apierrors.IsNotFound(err) {
		errs.Add(namespace, err)
		return warnings, errs, itemExists
	}
```

mpryc replied:
Should be fixed now.

sseago reviewed (commented):
The updated workflow is a lot clearer -- only one patch attempt each time the func is called. The only problems I see now are 1) handling of the obj we're adding managed fields for prior to generating patch, and 2) we shouldn't error our for NotFound errors, to replicate current behavior (where the NotFound logic was added a few releases ago to handle an edge case that a user ran into). Comments inline.

sseago commented on pkg/restore/restore.go:
```diff
@@ -1669,24 +1670,48 @@ func (ctx *restoreContext) restoreItem(obj *unstructured.Unstructured, groupReso
-			ctx.log.Infof("the managed fields for %s is patched", kube.NamespaceAndName(obj))
+			withoutManagedFields = createdObj.DeepCopy()
+		}
+
```
Add `withoutManagedFields = createdObj.DeepCopy()` here, right before calling setManagedFields.

mpryc replied:
Yeah that logic is much cleaner. thanks.

sseago commented on pkg/restore/restore.go:
```diff
@@ -1669,24 +1670,48 @@ func (ctx *restoreContext) restoreItem(obj *unstructured.Unstructured, groupReso
+		var patchErr error
+
+		if !firstPatchTry {
+			withoutManagedFields, patchErr = resourceClient.Get(name, metav1.GetOptions{})
```
s/withoutManagedFields/createdObj/

Commit:
fix: ARM images (#332) (#335)

* fix: ARM images

Signed-off-by: Mateus Oliveira <msouzaol@redhat.com>

* fixup! fix: ARM images

Signed-off-by: Mateus Oliveira <msouzaol@redhat.com>

---------

Signed-off-by: Mateus Oliveira <msouzaol@redhat.com>
(cherry picked from commit 41ad312a84cd70145cfd88aaa63ba734359c646a)

mpryc committed:
OADP-4640: Downstream only to allow override kopia default algorithms (#334)

Introduction of downstream only option to override Kopia default:
 - hashing algorithm
 - splitting algorithm
 - encryption algorithm

With introduction of 3 environment variables it is possible to override
Kopia algorithms used by Velero:

KOPIA_HASHING_ALGORITHM
KOPIA_SPLITTER_ALGORITHM
KOPIA_ENCRYPTION_ALGORITHM

If the env algorithms are not set or they are not within
Kopia SupportedAlgorithms, the default algorithm will be used.
This behavior is consistent with current behavior without this
change.

Signed-off-by: Michal Pryc <mpryc@redhat.com>

shubham-pampattiwar committed:
add missing unit test for kopia hashing algo (#337)

mpryc committed:
Further updates to the logic for properly restoring managed fields

Additional changes to the logic for restoring managed fields.

Signed-off-by: Michal Pryc <mpryc@redhat.com>
